### PR TITLE
[Rust] [DataFusion] [Experiment] Blocking threads filter

### PR DESCRIPTION
**Pull request by jorgecarleitao**
The motivation behind this PR is that Tokio does not really like blocking (e.g. cpu-intensive) operations, because (by definition) it forbids the scheduler from switching to other tasks. Because of this, tokio recommends (throughout its documentation, but most notably [here](https://docs.rs/tokio/1.0.1/tokio/index.html#cpu-bound-tasks-and-blocking-code)) to use `spawn_blocking` or `rayon` to handle blocking tasks.

This PR is just an experiment / proposal / idea of how we could handle this within tokio. Specifically, it is using `spawn_blocking` to spawn a thread on tokio's "blocking-dedicated" thread pool to handle a blocking operation, thereby avoiding starving the "async-dedicated" thread pool.

I do not expect this code to have much difference in performance, as `filter` is not such a blocking operation compared to e.g. a group by. However, I think that this idea could address performance issues when we have multiple stages (as one stage currently blocks the whole thread pool due to how we perform blocking ops inside `async` code).

@andygrove @alamb @Dandandan , I have been looking at DataFusion's code and tokio's documentation, and I hypothesize that this is would be one way to follow tokio's recommendations for our use-case, but I would really like to get your opinions.

### CODE REVIEW

File: rust/datafusion/src/physical_plan/sized_stream.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+use arrow::{datatypes::SchemaRef, error::Result as ArrowResult};
+use futures::stream::{Stream, StreamExt};
+
+use super::RecordBatchStream;
+
+/// A schema-aware [Stream] of [`RecordBatch`]es.
+pub struct SizedRecordBatchStream<S: Stream<Item = ArrowResult<RecordBatch>>> {
+    stream: Pin<Box<S>>,
+    schema: SchemaRef,
+}
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/sized_stream.rs:35]: This is just a way of encapsulating a stream into something that implements `RecordBatchStream`. Not really relevant for this PR's idea, just simplifies some common code.

File: rust/datafusion/src/physical_plan/filter.rs [hunk 2]
```diff
@@ -103,25 +103,23 @@ impl ExecutionPlan for FilterExec {
...
+            let predicate = predicate.clone();
+            async move {
+                // Filtering batches is CPU-bounded and therefore justifies a dedicated thread pool
+                task::spawn_blocking(move || batch_filter(&batch?, &predicate))
+                    .await
+                    .unwrap()
+            }
+        });
+
+        Ok(Box::pin(SizedRecordBatchStream::new(stream, self.schema())))
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/filter.rs:119]: This is the core idea: wrap the input stream with a blocking thread that runs `batch_filter` (the offending blocking code).

Comment (Dandandan) [reply rust/datafusion/src/physical_plan/filter.rs:119]: Does this code run multiple `batch_filter` for different batches at the same time with this code?

**Comment by Dandandan** [comment]
@jorgecarleitao 

This is really cool, thanks for creating this experiment!
I am not very deep yet into the Rust way of doing parallelism, the documentation of tokio makes sense to me.

Some ideas:

* In general, I think it is best if the parallelism is on a high level as possible to reduce the amount of overhead related to scheduling / context switching, etc.
* But in order to utilize parallelism best it should be fine-grained enough.
* I think there is some balance between total control of large amount of control control over parallelism. I think Spark concurrency via partitions is an example where you can have a larger amount of control over it. It is not always fine-grained enough, e.g. if you have one 1  / a couple of files as input.
* I think filtering batches is relatively fine-grained, so I am wondering if this a good level for parallelism.

* Tokios default config `max_blocking_threads` is 512, this is I think very large for CPU intensive work (and would have a negative effect on performance) https://docs.rs/tokio/1.0.1/tokio/runtime/struct.Builder.html#method.max_blocking_threads. Maybe if using different "scopes" it makes sense to use a different runtime for CPU-intensive work where you use a different `max_blocking_threads` config?
* Tokio's documentation seems to hint that Rayon would be a better choice for CPU intensive work?
* In the `ParquetExec` `thread::spawn` is being used. `task::spawn_blocking` seems a better choice there as it handles errors in a better way and can limit the nr. of threads compared to  thread::spawn` I guess?

* I think just as the statistics @andygrove started to add for `Exec`s it would be good to have something here as well to debug issues and make sure we are not doing things in an inefficient way

File: rust/datafusion/src/physical_plan/filter.rs (hunk 2 above)
Comment (Dandandan) [reply rust/datafusion/src/physical_plan/filter.rs:119]: It looks to me it currently does not?

**Review by alamb — COMMENTED** [review]
TLDR; It is my opinion that we do not (yet) have a problem that needs to be fixed.

I think this all comes down to what "CPU intensive" work mean.

At its core I would describe tokio / async as a cooperative multithreading scheme -- meaning that code needs to "yield" to a user level scheduler (aka the tokio runtime) "frequently" so other tasks are not starved. I realize the main tokio use case is async IO but that is just an input to the tokio scheduler in my mind.

In Rust these "yields" mean calls to `.await`

So in my mental model, what is required to use the tokio scheduler for "cpu intensive" work in datafusion is:
1. Call `.await` frequently enough that all other waiting tasks get a chance to run
2. Don't schedule so many CPU bound that some tasks that expect to run with low latency are starved.

I think using async streams satisfies 1 (as each record batch produced effectively `awaits` and I think gives tokio a chance to potentially process something new). 

I think 2 is challenging as tokio doesn't necessarily give any fairness guarantees across tasks or types of tasks and so as we get more and more complicated plans, we may increase the pressure on the scheduler and prevent IO or other operations from happening as quickly as they want. 

So as long as DataFusion operators operate on relatively small record batches and yield (aka call `.await` to get more input or produce output frequently) I don't think they fall into the "CPU intensive" category.

However I realize this is not a uniformly held opinion

The other alternative is to punt the non IO tasks scheduling problem to some other task queue / thread pool, as tokio suggests.  The upside is that it can be more carefully controlled and not interact with your IO tasks, but the downside is that you have to implement streaming / IO yourself and have now carved your resources (threads) up across pools and have to be careful about which pool you run what kind of work on

https://tokio.rs/blog/2020-04-preemption has some good background.

File: rust/datafusion/src/physical_plan/filter.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/filter.rs:119]: no, but this did not happen in master either: a stream does not ask for the next item until the current item is completed. We get no parallelism whatsoever from streams, only concurrency (because they yield). E.g. https://docs.rs/parallel-stream/2.1.2/parallel_stream/#differences-with-rayon

Comment (alamb) [reply rust/datafusion/src/physical_plan/filter.rs:119]: I think buffering with a channel is the Rust-y  way to get more parallel execution without unlimited buffering

Comment (andygrove) [reply rust/datafusion/src/physical_plan/filter.rs:119]: Thanks @jorgecarleitao this seems to make a lot of sense.

@alamb Was your comment a response to @Dandandan or an objection to using `spawn_blocking`? I wasn't sure.

Comment (alamb) [reply rust/datafusion/src/physical_plan/filter.rs:119]: My comment regarding channels was a response to @Dandandan  and @jorgecarleitao 's discussion of potentially getting more parallelism going. My comments regarding the threadpools was in response to this PR in general

**Comment by alamb** [comment]
@jorgecarleitao 
What is the status of this PR?

As part of trying to clean up the backlog of Rust PRs in this repo, I am going  through seemingly stale PRs and pinging the authors to see if there are any plans to continue the work or conversation.

**Comment by alamb** [comment]
@jorgecarleitao  I am closing this PR for the time being to clean up the Rust/Arrow PR backlog.  Please let me know if this is a mistake
